### PR TITLE
Migrate contributable menu locations from npe2

### LINKED ISSUE #5079: Define allowed plugin contribution
Currently, the list of menu contribution points are not defined, and we need to 
* figure out the strategy of whether we want to start with more constraint or less constraint (black list vs. white list)
* define the rules of the contribution point

more context https://github.com/napari/npe2/pull/161/files

current code: https://github.com/napari/napari/blob/e6c5576817c62c7900b27c577070d9e810b24758/napari/_app_model/constants/_menus.py#L38-L40

the output should be a miniNap or NAP documenting all the contribution points if cannot get consensus in quick discussions

### PR DESCRIPTION
**Pull request by liu-ziyang**
# Description
Migrate plugin contribution menu entries from npe2 PR: https://github.com/napari/npe2/pull/161

see 
Screen Shot 2022-10-21 at 9 20 11 AM

closes #5079 

## Type of change
- new feature

# References
#5079

### CODE REVIEW

**Review by kne42 — APPROVED** [review]
LGTM, failing tests appear to be caused by the same unrelated bugs affecting other PRs

**Comment by liu-ziyang** [comment]
@haesleinhuepf if you can help take a look here would be great, the list comes from the npe2 PR that you contributed to

File: napari/_app_model/_submenus.py
Comment (haesleinhuepf) [inline napari/_app_model/_submenus.py]: If the order is relevant, I would put Utilities last

**Review by haesleinhuepf — APPROVED** [review]
LGTM. Just a minor request / question

File: napari/_app_model/_submenus.py
Comment (liu-ziyang) [reply napari/_app_model/_submenus.py]: done!

File: napari/_app_model/constants/_menus.py
Comment (Czaki) [inline napari/_app_model/constants/_menus.py]: does this set need to be manually maintained, or maybe could be automatically generated from `MenuId.__members__`?

File: napari/_app_model/_submenus.py
Comment (Czaki) [inline napari/_app_model/_submenus.py]: maybe this also could be autogenerated. I strongly prefer to have a single place to edit not accidentally to introduce cone inconsistency.

**Review by Czaki — COMMENTED** [review]
It is possible to ensure code consistency by switching to autogenerate list instead of manually curating?

File: napari/_app_model/constants/_menus.py
Comment (liu-ziyang) [reply napari/_app_model/constants/_menus.py]: This needs to be manually managed. The contributable ones are not the full list of all menus

File: napari/_app_model/_submenus.py
Comment (liu-ziyang) [reply napari/_app_model/_submenus.py]: The submenu defines hierarchy of the menu structure and is therefore business logic and require manual management. It can probably be somewhat autogenerated based on the id structure but afaik we don't have a fixed convention, and even if we do, I suspect we will have some outliers later

File: napari/_app_model/constants/_menus.py
Comment (Czaki) [reply napari/_app_model/constants/_menus.py]: I still prefer to add some class methods to Enum, instead of a separate positions. It will be one place in the context of application logic.

Comment (liu-ziyang) [reply napari/_app_model/constants/_menus.py]: I tried adding the contributable as a boolean to the enum and some class methods to check if a menu id is contributable or not. My biggest complaint is that we have to either use the private member map in the enum class or loop through all values to match the id in question. Either way the code doesn't look very nice. If you still prefer this approach, I can push what I have and see if you like them.

File: napari/_qt/_qapp_model/_tests/test_qapp_model_menus.py
Comment (DragaDoncila) [inline napari/_qt/_qapp_model/_tests/test_qapp_model_menus.py]: I would much prefer to see this test skipped explicitly for the tools menu which afaik is the only one that may reasonably have 0 actions. I'm also not sure what you mean by placeholder submenus? Why do they not show up under `app.menus`?

File: napari/_app_model/constants/_menus.py
Comment (DragaDoncila) [reply napari/_app_model/constants/_menus.py]: I don't super mind these being declared manually in each spot, I do think it's part of the necessary curation work that we'll need to do as we update these menus. Happy to be challenged on that.

I think if we're going to go with this whitelisting approach, we need a warning [here](https://github.com/napari/napari/blob/9ea0159ad2b690556fe56ce480886d8f0b79ffae/napari/plugins/_npe2.py#L353) that asks users to suggest that menu for whitelisting, and mentions the menus that are being skipped.

**Review by DragaDoncila — CHANGES_REQUESTED** [review]
I don't necessarily mind declaring these menus manually, I do think it's part of the necessary curation/upkeep we'll need to maintain, especially as the list expands. This seems to work as expected, and the logic for only including `_CONTRIBUTABLE` menu contributions is already present.

I left a comment on the test update, asking that it only specifically skips Tools menus. I'd also like it if we add an explicit warning for when someone tries to contribute to a menu that is not contributable.

File: napari/_app_model/constants/_menus.py
Comment (kevinyamauchi) [reply napari/_app_model/constants/_menus.py]: > I don't super mind these being declared manually in each spot, I do think it's part of the necessary curation work that we'll need to do as we update these menus. Happy to be challenged on that.

I also don't mind if there isn't a straightforward way to generate this list, but it might be nice to have a test to make sure these items are consistent with `MenuId`. Is that possible?

> I think if we're going to go with this whitelisting approach, we need a warning [here](https://github.com/napari/napari/blob/9ea0159ad2b690556fe56ce480886d8f0b79ffae/napari/plugins/_npe2.py#L353) that asks users to suggest that menu for whitelisting, and mentions the menus that are being skipped.

agreed!

**Review by kevinyamauchi — APPROVED** [review]
Thanks, @potating-potato ! I had a look at the menu categories and I think they look good!

I don't know the machinery around this well enough to review the code, so I hope others can comment.

File: napari/_app_model/constants/_menus.py
Comment (Czaki) [reply napari/_app_model/constants/_menus.py]: > I also don't mind if there isn't a straightforward way to generate this list, but it might be nice to have a test to make sure these items are consistent with `MenuId`. Is that possible?

The most common approach that I see is to define a class method for each category in enum to return a list of enum members belonging to a given category. Then you store such information in class, not in other files.

File: napari/_qt/_qapp_model/_tests/test_qapp_model_menus.py
Comment (liu-ziyang) [reply napari/_qt/_qapp_model/_tests/test_qapp_model_menus.py]: under app model, sub menus and menus are treated differently, they both have an ID that is declared under menu constants but they end up being created as different classes, which is why app_model.menus() do not show these, as menus are MenuItem stored under _menu_items in the app model, and submenus exists within MenuItem as SubMenuItem

I am fine changing the check to exclude tools menu

File: napari/_app_model/constants/_menus.py
Comment (liu-ziyang) [reply napari/_app_model/constants/_menus.py]: > > I also don't mind if there isn't a straightforward way to generate this list, but it might be nice to have a test to make sure these items are consistent with `MenuId`. Is that possible?
> 
> The most common approach that I see is to define a class method for each category in enum to return a list of enum members belonging to a given category. Then you store such information in class, not in other files.

That makes sense! I probably took auto-generation too literally. I can move the contributable list to the menuID and make it a class method

**Comment by liu-ziyang** [comment]
> I don't necessarily mind declaring these menus manually, I do think it's part of the necessary curation/upkeep we'll need to maintain, especially as the list expands. This seems to work as expected, and the logic for only including `_CONTRIBUTABLE` menu contributions is already present.
> 
> I left a comment on the test update, asking that it only specifically skips Tools menus. I'd also like it if we add an explicit warning for when someone tries to contribute to a menu that is not contributable.

@DragaDoncila I hope I addressed all of your comments, can you please take another look? thanks!

File: napari/_app_model/_submenus.py
Comment (Czaki) [reply napari/_app_model/_submenus.py]: maybe also this could be generated base on classmethod?

Comment (liu-ziyang) [reply napari/_app_model/_submenus.py]: Good call! I moved this as well to a classmethod

File: napari/_app_model/constants/_menus.py
Comment (Czaki) [inline napari/_app_model/constants/_menus.py:51]: docstring?

File: napari/_app_model/constants/_menus.py
Comment (Czaki) [inline napari/_app_model/constants/_menus.py:104]:
```suggestion
        menu_id in MenuId.contributables() or not menu_id.startswith("napari/")
```

File: napari/plugins/_npe2.py
Comment (Czaki) [inline napari/plugins/_npe2.py]: stacklevel? category?

**Review by Czaki — APPROVED** [review]
few comments. 
I have mixed feelings as I do not like deep menus with multiple entries. But as I understand many researchers are familiar. 

I prefer to have some possibility to search for utility, but maybe merge this and check what happens. We could always drop it and replace it with a "search tool widget".

File: napari/_app_model/constants/_menus.py
Comment (kne42) [reply napari/_app_model/constants/_menus.py:104]: id rather this be expanded to a full-blown if-else statement with comments rather than a confusing one-liner

Comment (Czaki) [reply napari/_app_model/constants/_menus.py:104]: ok

**Comment by liu-ziyang** [comment]
I got feedback that the menu structure would need a bigger discussion and potentially a NAP to record the decisions as it is influential. I removed the tools menu migrated from npe2 as a result and reduced this PR to only the refactor idea suggested. I will leave this to core dev's discretion whether this refactor PR should be merged/closed

**Comment by liu-ziyang** [comment]
closing this, superseded by NAP #6

**Pull request closed without merging**
